zerver/migrations/0386_fix_attachment_realm.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any
+
+from django.conf import settings
+from django.db import migrations
+from django.db.backends.postgresql.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+RECIPIENT_PERSONAL = 1
+RECIPIENT_STREAM = 2
+
+
+def fix_attachment_realm(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    Attachment = apps.get_model("zerver", "Attachment")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    Stream = apps.get_model("zerver", "Stream")
+
+    def get_realm_id_from_recipient(recipient: Any) -> Any:
+        if recipient.type == RECIPIENT_PERSONAL:
+            return UserProfile.objects.only("realm_id").get(id=recipient.type_id).realm_id
+
+        if recipient.type == RECIPIENT_STREAM:
+            return Stream.objects.only("realm_id").get(id=recipient.type_id).realm_id
+
+        raise AssertionError("Recipient is not PERSONAL or STREAM")
+
+    if not Realm.objects.exists():
+        return
+
+    internal_realm = Realm.objects.get(string_id=settings.SYSTEM_BOT_REALM)
+
+    print("")
+    for attachment in Attachment.objects.filter(realm_id=internal_realm.id).order_by("id"):
+        print(f"Fixing attachment {attachment.id}")
+        message = attachment.messages.only("recipient").first()
+        if message is None:
+            # If attachment.messages is empty, nothing to do.
+            continue
+
+        message_realm_id = get_realm_id_from_recipient(message.recipient)
+        attachment.realm_id = message_realm_id
+        attachment.save(update_fields=["realm_id"])
```

Comment:
Do we want to re-write this logic for proper bulk operations? I'm not certain if it's worth it though, because that's probably only needed for Zulip Cloud and a simpler migration strategy there might be to just add some `time.sleep()` here and let the code slowly run in the background? (could even just be done by copy-pasting into `manage.py shell` rather than using actually `manage.py migrate`)

Reply:
In Zulip Cloud, there are 27k messages that this touches.  The rows in `zerver_attachment` aren't that wide, so this is probably also OK from a memory point of view.

Reply:
Yeah, there's no need to do bulk operations here, since it's a relatively rare situation that created rows needing this attention.